src/pydot/dot_parser.py

```diff
@@ -410,17 +407,7 @@ def graph_definition():
 
         righthand_id = (float_number | ID).setName("righthand_id")
 
-        port_angle = (at + ID).setName("port_angle")
-
-        port_location = (
-            OneOrMore(Group(colon + ID))
-            | Group(colon + lparen + ID + comma + ID + rparen)
-        ).setName("port_location")
-
-        port = (
-            Group(port_location + Optional(port_angle))
-            | Group(port_angle + Optional(port_location))
-        ).setName("port")
+        port = Group(OneOrMore(Group(colon + ID))).setName("port")
```

Comment:
I wonder if this should be a `OneOrMore`, or if it should explicitly limit the number of colons to 2? IOW,

```python
port = (
    Group(":" + ID + ":" + ID)
    | Group(":" + ID)
).set_name("port")
```
(In that order because `|` alternation takes the first branch that matches.)

I'd like to `Suppress()` the leading colon, as well, but our current weird logic keeps it as part of the `obj_dict["port"]`. So I think it has to be captured here as well, until/unless we fix that.

Reply:
Noted for the future. 

I even changed it but it leads to test failures. Because of some logic in the `do_node_ports`, the `Group` wrapping has to be kept:
```python
port = Group(
    Group(":" + ID + ":" + ID)
    | Group(":" + ID)
).setName("port")
```

But even with that, a single test (`ports.dot`) is still failing. So that will need a logic change.

Reply:
Hmm, odd. I'll try to take a look at that, at some point.

As another mid-term thing, the parsing code can actually be improved a _lot_ if we switch to using results names.

Right now, all of the parse actions walk the token list picking out potentially interesting things to process. But there's a much easier way.

Say I have a `node_stmt` defined like this:
```python
node_stmt = (
    node_id.set_results_name("node_id")
    + Optional(attr_block).set_results_name("node_attrs")
    + Optional(';').suppress()
)
```
(Where `node_id` and `attr_block` are defined appropriately.)

Then, when I parse a node line, the relevant items will automatically be identified with the results names assigned, and can be accessed that way directly. (`node_stmt` in this example has no parse actions set on it, initially.)

```python
>>> from pprint import pp
>>> res = node_stmt.parse_string(
... "a [color=red, shape=box];")
>>> pp(res.as_dict())
{'node_id': ['a'],
 'node_attrs': [['color', 'red'], ['shape', 'box']]}
>>> res['node_id']
ParseResults(['a'], {})
```

...And they can be accessed that way from _inside_ the parse actions, as well.

```python
>>> def demo(toks):
...    _id = toks.get('node_id')
...    attrs = toks.get('node_attrs')
...    print(f"{_id}\n{attrs}")
...
>>> node_stmt.set_parse_action(demo)
node_stmt
>>> res = node_stmt.parse_string(
... "a [color=red, shape=box];")
['a']
[['color', 'red'], ['shape', 'box']]
>>> 
```

Used correctly, it's a much more deterministic way of consuming the parse results than just walking the raw token list. By doing that, the parse action is effectively re-parsing the input. A lot of pyparsing's power is going to waste, since it's just acting as an input validator and dumb tokenizer.

Reply:
I figured it out. Because of the assumptions about token hierarchy in `push_edge_stmt()` and `do_node_ports()`, it has to be:

```python
port = (
    Group(Group(colon + ID) + Group(colon + ID))
    | Group(Group(colon + ID))
).setName("port")
```

...But that does work. I'll put together a PR.